x-pack/plugin/kql/src/main/java/org/elasticsearch/xpack/kql/query/KqlQueryBuilder.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.kql.query;
+
+import org.apache.lucene.search.Query;
+import org.elasticsearch.TransportVersion;
+import org.elasticsearch.common.ParsingException;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.index.query.AbstractQueryBuilder;
+import org.elasticsearch.index.query.QueryBuilder;
+import org.elasticsearch.index.query.SearchExecutionContext;
+import org.elasticsearch.logging.LogManager;
+import org.elasticsearch.logging.Logger;
+import org.elasticsearch.xcontent.ConstructingObjectParser;
+import org.elasticsearch.xcontent.ParseField;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentParser;
+import org.elasticsearch.xpack.kql.parser.KqlParser;
+
+import java.io.IOException;
+import java.util.Objects;
+
+import static org.elasticsearch.xcontent.ConstructingObjectParser.constructorArg;
+
+public class KqlQueryBuilder extends AbstractQueryBuilder<KqlQueryBuilder> {
+    public static final String NAME = "kql";
+    public static final ParseField QUERY_FIELD = new ParseField("query");
+    private static final Logger log = LogManager.getLogger(KqlQueryBuilder.class);
+    private static final ConstructingObjectParser<KqlQueryBuilder, Void> PARSER = new ConstructingObjectParser<>(
+        NAME,
+        a -> new KqlQueryBuilder((String) a[0])
+    );
+
+    static {
+        PARSER.declareString(constructorArg(), QUERY_FIELD);
+        declareStandardFields(PARSER);
+    }
+
+    private final String query;
+
+    public KqlQueryBuilder(String query) {
+        this.query = Objects.requireNonNull(query, "query can not be null");
+    }
+
+    public KqlQueryBuilder(StreamInput in) throws IOException {
+        super(in);
+        query = in.readString();
+    }
+
+    public static KqlQueryBuilder fromXContent(XContentParser parser) {
+        try {
+            return PARSER.apply(parser, null);
+        } catch (IllegalArgumentException e) {
+            throw new ParsingException(parser.getTokenLocation(), e.getMessage(), e);
+        }
+    }
+
+    @Override
+    protected void doXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject(NAME);
+        builder.field(QUERY_FIELD.getPreferredName(), query);
+        boostAndQueryNameToXContent(builder);
+        builder.endObject();
+    }
+
+    @Override
+    protected QueryBuilder doSearchRewrite(SearchExecutionContext searchExecutionContext) throws IOException {
```

Comment:
I am surprised that our rewrite requires the full search rewrite on the data node. I would have assumed, since kql is parsable and usable in Kibana, we could rewrite much earlier into a set of QueryBuilder objects and serialize those down to the data nodes.

Reply:
Looking at `parseKqlQuery`, I might be missing something, but all you need is the `mappingLookup` right?

It is interesting how this parser needs "more access" than the Kibana one. 

Being able to rewrite on the coordinator allows us to take advantage of powerful coordinator optimizations.

However, if this is NOT possible, we should still strive to use (at a minimum) to rewrite with `convertToIndexMetadataContext`, while this is still done on the data node (thus cannot take advantage of coordinator skipping), it will allow this to be rewritten during the `can_match` phase. 

NOTE: If we cannot rewrite during coordinator, I would expect the performance of this query to be measurably worse than the Kibana parser at scale (e.g. with many searchable snapshots, etc.).

Reply:
KqlParser does indeed need field mapping access since it builds queries conditionnally on field types. It’s not entirely true to say it requires more access than Kibana, it also accesses mapping info through the dataview abstraction and build mapping from them.

Given that we need an access to the mapping, I do not think it is possible to move the rewriting on the coordinator node. I have done some refactoring to move it into the doIndexMetadataRewrite method of KqlQueryBuilder as suggested.

Performance-wise, you make a great point.Generally speaking, performance should be similar to query_string, which also does parsing on the data node.

Our main goal is to expose this in ES|QL for an easy migration path, so if query_string works well enough, KQL should too!

Reply:
Thank you for switching to use `doIndexMetadataRewrite` this allows for skipping in `can_match` on the data node. Which greatly improves things.